go/pkg/bertymessenger/handlers.go

```diff
@@ -304,536 +439,398 @@ func groupMemberDeviceAdded(svc *service, gme *bertytypes.GroupMetadataEvent) er
-						Error
-				}
+	member, err := h.db.addMember(i.MemberPublicKey, i.ConversationPublicKey, payload.GetName())
+	if err != nil {
+		return i, err
+	}
 
-				member := Member{
-					PublicKey:             i.MemberPublicKey,
-					ConversationPublicKey: gpk,
-					DisplayName:           payload.GetName(),
-				}
-				err := tx.
-					Clauses(clause.OnConflict{
-						Columns:   []clause.Column{{Name: "public_key"}},
-						DoUpdates: clause.AssignmentColumns([]string{"display_name"}),
-					}).
-					Create(&member).
-					Error
-				if err != nil {
-					return err
-				}
+	if h.svc != nil {
+		err = h.svc.dispatcher.StreamEvent(StreamEvent_TypeMemberUpdated, &StreamEvent_MemberUpdated{Member: member})
+		if err != nil {
+			return i, err
+		}
 
-				err = svc.dispatcher.StreamEvent(StreamEvent_TypeMemberUpdated, &StreamEvent_MemberUpdated{Member: &member})
-				if err != nil {
-					return err
-				}
+		h.logger.Debug("dispatched member update", zap.String("name", payload.GetName()), zap.String("device-pk", i.GetDevicePublicKey()), zap.String("conv", i.ConversationPublicKey))
+	}
 
-				svc.logger.Debug("dispatched member update", zap.String("name", payload.GetName()), zap.String("device-pk", i.GetDevicePublicKey()), zap.String("conv", gpk))
+	return i, nil
+}
 
-			default:
-				svc.logger.Warn("unsupported app message type", zap.String("type", i.GetType().String()))
-			}
+func interactionFromAppMessage(h *eventHandler, gpk string, gme *bertytypes.GroupMessageEvent, am *AppMessage) (*Interaction, error) {
+	amt := am.GetType()
+	cidb := gme.GetEventContext().GetID()
+	cid := bytesToString(cidb)
+	isMe, err := checkIsMe(h.ctx, h.protocolClient, gme)
+	if err != nil {
+		return nil, err
+	}
 
-			if isVisibleEvent {
-				// FIXME: check if app is in foreground
+	dpkb := gme.GetHeaders().GetDevicePK()
+	dpk := bytesToString(dpkb)
 
-				// fetch conversation from db
-				conv, err := getConversationByPK(tx, conv.GetPublicKey())
-				if err != nil {
-					return err
-				}
+	i := &Interaction{
+		CID:                   cid,
+		Type:                  amt,
+		Payload:               am.GetPayload(),
+		IsMe:                  isMe,
+		ConversationPublicKey: gpk,
+		SentDate:              am.GetSentDate(),
+		DevicePublicKey:       dpk,
+	}
+	h.logger.Debug("received app message", zap.String("type", amt.String()))
 
-				// visible events update the last_update field
-				updates := map[string]interface{}{
-					"last_update": timestampMs(time.Now()),
-				}
+	return i, nil
+}
 
-				// if conv is not open, increment the unread_count
-				if !i.IsMe && !conv.GetIsOpen() {
-					updates["unread_count"] = gorm.Expr("unread_count + 1")
-				}
+func (h *eventHandler) hydrateInteraction(gme *bertytypes.GroupMessageEvent, i *Interaction) error {
+	// fetch conv from db
+	conversation, err := h.db.getConversationByPK(i.ConversationPublicKey)
+	if err != nil {
+		// FIXME: maybe we should accept receiving app messages on unknown conversations?
```

Comment:
yes, I think we should keep the warn but not return an error, I'm curious about when it will happen